### ISSUE TITLE
an example of how to use nuklear to make gui?

### ISSUE THREAD

**Issue by litlight** [ytsarko/nuklear-android-ndk-sample#1]
how to use nuklear to display the gui? should i use c code or java code?

**Comment by ytsarko** [comment]
Hi @litlight,

nuklear is 1 header file C library. It is much easier to use C language to use it. In the case of Java you will have to create special C library (google for JNI) and expose interfaces of nuklear into Java world. This repository contains a very simplified, high-level example of how one can expose some functions into Java world with implementation done in C.

If you need only to expose nuklear API to Java you will need create a class representing nuklear with methods corresponding to its drawing APIs (the simplest and straightforward solution) or break down the functionality of nuklear into a number of Java classes grouped by some function they are doing (primitive drawing, widgets, etc) which the user of the library can later consume.

Currently I don't have plans on providing such complete JNI library.

**Comment by litlight** [comment]
@ytsarko Thank you, finally I can run the official nuklear demo on Android with allegro5 backend. I just wonder with the opengl backend in this example, can i run the nuklear demo directly? in the official nuklear demo, to show the nuklear gui, we need to call nk_input_begin, nk_xxx_handle_event, nk_begin_xxx, nk_xxx_render in a while loop, can we bind a c side main loop function with the opengl backend in this example to run the official demo?

**Comment by ytsarko** [comment]
@litlight, this repository only contains demo of how to create simple GL view in Android app and provide a special renderer for it that, in turn, should use calls to nk_gl_render to do actual nuklear rendering.

So to make it more functional you need to call nk_gl_init in Java_com_example_yt_myapplication_LibraryClass_init, nk_gl_render in static render(w,h).

This will render whatever you've asked nuklear for using its rendering APIs (nk_xxx_yyy).

Of course in order to make it more fun you'll have to intercept user actions like taps and finger moves using Android APIs and forward those to nuklear using nk_input_begin/end and nk_input_motion/key/button and etc depending on the incoming action.

This all is left out of this simple demo scope. I suggest you take some more recent nuklear+opengl sample from github and port it over/adapt it here. I wasn't following the nuklear for some years, so code in this repo might be quite obsolete.

**Comment by ytsarko** [comment]
In addition to my previous post, I really strongly encourage to port a modern nuklear+opengl example into this demo, because I don't really remember the current implementation of nk_gl_device_create - why it calls setup_null_texture and setup_framebuffer. Perhaps I've used wrong implementation when creating this demo project. At the time of this demo creation I was playing with different backends for nuklear: framebuffer, opengl, directx, cairo to name a few, so perhaps I've used some other backend here instead of one proper for opengl.
